src/main/java/org/prebid/server/bidder/criteo/CriteoBidder.java

```diff
@@ -0,0 +1,293 @@
+                sizes.add(String.format("%sx%s", format.getW(), format.getH()));
+            }
+        } else if (banner.getW() != null && banner.getW() > 0
+                && banner.getH() != null && banner.getH() > 0) {
+            sizes.add(String.format("%sx%s", banner.getW(), banner.getH()));
+        }
+        return sizes;
+    }
+
+    private ExtImpCriteo parseImpExt(Imp imp) {
+        try {
+            return jsonMapper.mapper().convertValue(imp.getExt().get("bidder"), ExtImpCriteo.class);
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+    }
+
+    private static boolean allSlotsAreContainingSameNetworkId(List<CriteoRequestSlot> criteoRequestSlots) {
+        return criteoRequestSlots.stream()
+                .map(CriteoRequestSlot::getNetworkId)
+                .filter(Objects::nonNull)
+                .distinct()
+                .count() > 1;
+    }
+
+    private static ExtRegs getExtRegs(BidRequest bidRequest) {
+        final Regs regs = bidRequest.getRegs();
+        return regs != null ? regs.getExt() : null;
+    }
+
+    private static CriteoPublisher buildCriteoPublisher(BidRequest bidRequest, Integer networkId) {
+        final CriteoPublisher.CriteoPublisherBuilder criteoPublisherBuilder = CriteoPublisher.builder();
+        if (networkId != null) {
+            criteoPublisherBuilder.networkId(networkId);
+        }
+
+        if (bidRequest.getApp() != null) {
+            criteoPublisherBuilder.bundleId(bidRequest.getApp().getBundle());
+        }
+
+        if (bidRequest.getSite() != null) {
+            criteoPublisherBuilder
+                    .siteId(bidRequest.getSite().getId())
+                    .url(bidRequest.getSite().getPage());
+        }
+
+        return criteoPublisherBuilder.build();
+    }
+
+    private static CriteoUser buildCriteoUser(BidRequest bidRequest, ExtRegs extRegs) {
+        final CriteoUser.CriteoUserBuilder criteoUserBuilder = CriteoUser.builder();
+        if (bidRequest.getUser() != null) {
+            criteoUserBuilder.cookieId(bidRequest.getUser().getBuyeruid());
+        }
+
+        if (bidRequest.getDevice() != null) {
+            criteoUserBuilder
+                    .deviceIdType(determineDeviceIdType(bidRequest.getDevice().getOs()))
+                    .deviceOs(bidRequest.getDevice().getOs())
+                    .deviceId(bidRequest.getDevice().getIfa())
+                    .ip(bidRequest.getDevice().getIp())
+                    .ipV6(bidRequest.getDevice().getIpv6())
+                    .userAgent(bidRequest.getDevice().getUa());
+        }
+
+        if (extRegs != null) {
+            criteoUserBuilder.uspIab(extRegs.getUsPrivacy());
+        }
+
+        return criteoUserBuilder.build();
+    }
+
+    private static String determineDeviceIdType(String deviceOs) {
+        if (deviceOs == null) {
+            return "unknown";
+        }
+        switch (deviceOs.toLowerCase()) {
```

Comment:
We can make var like `final lowerCaseDeviceOs = StringUtils.stripToEmpty(deviceOs).toLowerCase()` and pass to `switch` statemant.

Reply:
Done